File: go/0572/0572_SubtreeofAnotherTree.go
// 572. Subtree of Another Tree

package main

import (
	"bufio"
	"fmt"
	"leetcode/structs"
	"leetcode/utils"
	"os"

	"github.com/samber/lo"
)

func isSameTree(left *structs.TreeNode, right *structs.TreeNode) bool {
	if left == nil && right == nil {
		return true
	}
	if left == nil || right == nil || left.Val != right.Val {
		return false
	}
	return isSameTree(left.Left, right.Left) && isSameTree(left.Right, right.Right)
}

func isSubtree(root *structs.TreeNode, subRoot *structs.TreeNode) bool {
	if root == nil {
		return false
	}
	if root.Val == subRoot.Val && isSameTree(root, subRoot) {
		return true
	}
	return isSubtree(root.Left, subRoot) || isSubtree(root.Right, subRoot)
}

func main() {
	scanner := bufio.NewScanner(os.Stdin)
	var lines []string
	for scanner.Scan() {
		lines = append(lines, scanner.Text())
	}

	const testSize int = 2

	for _, test := range lo.Chunk(lines, testSize) {
		root := utils.ToTree(test[0])
		subRoot := utils.ToTree(test[1])
		result := isSubtree(root, subRoot)
		fmt.Println(result)
	}
}

File: py/0572_SubtreeofAnotherTree.py
# 572. Subtree of Another Tree

from utils import chunk, toTree, printTree
from typing import Optional
from classes import TreeNode

# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right


class Solution:
    def isSameTree(self, p: Optional[TreeNode], q: Optional[TreeNode]) -> bool:
        if not p and not q:
            return True
        if not p or not q:
            return False
        return p.val == q.val and self.isSameTree(p.left, q.left) and self.isSameTree(p.right, q.right)

    def isSubtree(self, root: Optional[TreeNode], subRoot: Optional[TreeNode]) -> bool:
        if not root:
            return False
        return root and self.isSameTree(root, subRoot) or self.isSubtree(root.left, subRoot) or self.isSubtree(root.right, subRoot)


if __name__ == "__main__":
    testSize = 2
    puzzles = [
        [3, 4, 5, 1, 2],
        [4, 1, 2],
        [3, 4, 5, 1, 2, None, None, None, None, 0],
        [4, 1, 2],
    ]
    for puzzle in chunk(puzzles, testSize):
        print(Solution().isSubtree(*map(toTree, puzzle)))
